File: fresh/installer.py
import shlex
import subprocess


class Installer:
    @classmethod
    def install_app(cls, spinner, app, installer):
        app_name = app
        if installer == "brew":
            install_command = f"brew install {app}"
        elif installer == "cask":
            install_command = f"brew install --cask {app}"
        elif installer == "mas":
            install_command = f"mas install {app['id']}"
            app_name = app["name"]
        else:
            raise ValueError("Invalid installer")
        spinner.start(app_name)
        args = shlex.split(install_command)
        try:
            proc = subprocess.Popen(
                args, stdout=subprocess.PIPE, stderr=subprocess.PIPE
            )
        except OSError as e:
            spinner.fail()
            print(f"Error running command: {install_command}")
            print(e)
        _, stderr = proc.communicate()
        if proc.returncode == 0:
            spinner.succeed()
        else:
            spinner.fail()
            print(stderr.decode())

File: src/installer.js
import { run } from "./util/exec.js";

// Install a single application via brew, brew --cask, or mas.
// `app` is either a string (brew/cask) or { id, name } (mas).
export async function installApp(spinner, app, kind) {
  let displayName;
  let cmd;
  let args;

  switch (kind) {
    case "brew":
      displayName = app;
      cmd = "brew";
      args = ["install", app];
      break;
    case "cask":
      displayName = app;
      cmd = "brew";
      args = ["install", "--cask", app];
      break;
    case "mas":
      displayName = app.name;
      cmd = "mas";
      args = ["install", String(app.id)];
      break;
    default:
      throw new Error(`Invalid installer kind: ${kind}`);
  }

  spinner.start(displayName);
  try {
    const result = await run(cmd, args);
    if (result.exitCode === 0) {
      spinner.succeed(displayName);
    } else {
      spinner.fail(displayName);
      if (result.stderr) console.error(result.stderr);
    }
  } catch (err) {
    spinner.fail(displayName);
    console.error(`Error running ${cmd} ${args.join(" ")}`);
    console.error(err?.message ?? err);
  }
}
